adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java

```diff
@@ -438,64 +443,30 @@ private void handleConnectionRequestResult(final MqttEndpoint endpoint,
 
             } else {
-
-                return getTenantConfiguration(credentials.getTenantId(), currentSpan.context()).compose(tenantConfig -> {
-                    if (tenantConfig.isAdapterEnabled(getTypeName())) {
-                        LOG.debug("protocol adapter [{}] is enabled for tenant [{}]",
-                                getTypeName(), credentials.getTenantId());
-                        return Future.succeededFuture(tenantConfig);
-                    } else {
-                        LOG.debug("protocol adapter [{}] is disabled for tenant [{}]",
-                                getTypeName(), credentials.getTenantId());
-                        return Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_FORBIDDEN,
-                                "adapter disabled for tenant"));
-                    }
-                }).compose(tenantConfig -> {
-                    final Future<DeviceUser> result = Future.future();
-                    usernamePasswordAuthProvider.authenticate(credentials, result.completer());
-                    return result;
-                }).compose(authenticatedDevice -> {
-                    currentSpan.log(String.format("device authenticated"));
-                    LOG.debug("successfully authenticated device [tenant-id: {}, auth-id: {}, device-id: {}]",
-                            authenticatedDevice.getTenantId(), credentials.getAuthId(),
-                            authenticatedDevice.getDeviceId());
-                    return triggerLinkCreation(authenticatedDevice.getTenantId()).map(done -> {
-                        currentSpan.log(String.format("opened downstream links"));
-                        onAuthenticationSuccess(endpoint, authenticatedDevice);
-                        return null;
-                    }).compose(ok -> accepted(authenticatedDevice));
-                }).recover(t -> {
-                    LOG.debug("cannot establish connection with device [tenant-id: {}, auth-id: {}]",
-                            credentials.getTenantId(), credentials.getAuthId(), t);
-                    if (t instanceof ServerErrorException) {
-                        // one of the services we depend on might not be available (yet)
-                        return rejected(MqttConnectReturnCode.CONNECTION_REFUSED_SERVER_UNAVAILABLE);
-                    } else {
-                        // validation of credentials has failed
-                        return rejected(MqttConnectReturnCode.CONNECTION_REFUSED_NOT_AUTHORIZED);
-                    }
-                });
-
+                return getTenantConfiguration(credentials.getTenantId(), currentSpan.context())
+                        .compose(tenantConfig -> isAdapterEnabled(credentials, tenantConfig))
+                        .compose(tenantConfig -> authenticateCredentials(credentials))
+                        .compose(authenticatedDevice -> isDeviceRegistered(authenticatedDevice, credentials,
```

Comment:
running all of these checks sequentially will increase latency (a lot). IMHO we should retrieve all relevant data and make the checks in parallel, failing the (composite) future as soon as one of the checks fails ...

Reply:
FMPOV `isAdapterEnabled()` and `authenticateCredentials()` can be checked in parallel and the code snippet is as below.  And the rest have dependency on the result of another future and I don't see many options for composite future. Technically it is possible to execute `createLink()` in parallel with the rest. FMPOV why should we even create links until all the checks are successful.
```
final Future<DeviceUser> authenticatedDeviceTracker = authenticateCredentials(credentials);
                
getTenantConfiguration(credentials.getTenantId(), currentSpan.context())
     .compose(tenantObject -> CompositeFuture.all(isAdapterEnabled(tenantObject), authenticatedDeviceTracker))
     .compose(x->isDeviceRegistered(authenticatedDeviceTracker.result(), credentials, currentSpan))
     .compose(authenticatedDevice->createLink(endpoint, authenticatedDevice, currentSpan))
```

Reply:
The point is that the *retrieval* of the data is very expensive (i.e. slow) compared to the logical checks we need to perform. So, instead of doing it sequentially, we can at least *retrieve* the data in parallel
* tenant
* credentials
* device registration
If any of those retrieval operations fails, then the device cannot connect. Otherwise, we need to do the logical checks (which are cheap to perform), i.e.
* is tenant enabled?
* are the credentials correct?
* is the device enabled?

Sometimes we might not actually validate the retrieved registration information anymore because the tenant has already been determined to be disabled, but that is no issue FMPOV.

Reply:
I fully agree with it with the benefits of _retrieval_ of data  in parallel.  IMHO, at least the credentials should be retrieved and verified first before retrieving the rest (tenant or device registration). If we retrieve tenants, devices etc even before verifying credentials, then in case of _invalid_ credentials or any _attack_, it adds unnecessary overload to the system or prone to attack.

Reply:
sounds like a plan